plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java

```diff
@@ -240,16 +238,15 @@
     private final AtomicReference<Scheduler.Cancellable> finalizationFuture = new AtomicReference<>();
 
     @Override
-    public void finalizeSnapshot(SnapshotId snapshotId, ShardGenerations shardGenerations, long startTime, String failure, int totalShards,
-                                 List<SnapshotShardFailure> shardFailures, long repositoryStateId, boolean includeGlobalState,
-                                 Metadata clusterMetadata, Map<String, Object> userMetadata, Version repositoryMetaVersion,
+    public void finalizeSnapshot(SnapshotId snapshotId, ShardGenerations shardGenerations, long repositoryStateId, Metadata clusterMetadata,
+                                 Supplier<SnapshotInfo> buildSnapshotInfo, Version repositoryMetaVersion,
```

Comment:
Are we using a `Supplier` here so that the `SnapshotInfo` end time is updated when it is built?

Reply:
Exactly that's why. Now that you're asking I wonder if that's even necessary ... the actual finalisation step on the repository is super parallelised these days anyway and will take a trivial amount of time. Could also just pass in the full `SnapshotInfo` I guess ... but then again, it gets all weird if someone actually finds a way of creating a situation where finalisation is slow (huge number of indices could maybe cause this).
... I think for now I like this best because it doesn't change behaviour :)

Reply:
What do you think of passing a SnapshotInfo instance to the finalizeSnapshot() but update the end time of this one right before it's written to the repository?

```java
executor.execute(ActionRunnable.supply(allMetaListener, () -> {
                final SnapshotInfo finalSnapshotInfo = SnapshotInfoBuilder.builder(snapshotInfo)
                                .withEndTime(threadPool.absoluteTimeInMillis())
                                .build();
                snapshotFormat.write(snapshotInfo, blobContainer(), snapshotId.getUUID(), false);
                return snapshotInfo;
            }));
```

Reply:
Hmm, then we still have some of the responsibility for setting up the correct `SnapshotInfo` in the blob store and repository implementations in general which I tried to avoid. **But :)**
I just realized that my above comment doesn't make 100% sense these days ...  we are writing the `SnapshotInfo` concurrently to index metadata and global metadata blobs with no defined order so the timestamp in it isn't really 100% correct in the first place. And, even if we were to write it last it wouldn't be because we still have to do two CS updates and the write of the new `index-N` in between those afterwards.
=> how about we just pass `SnapshotInfo` and not care about the precision of this timestamp since it can never be 100% correct anyway?

Reply:
Sold!

Reply:
Alrighty did this now :) Constructing `SnapshotInfo` in the snapshots service simplified things massively it turns out. See https://github.com/elastic/elasticsearch/pull/58834/commits/123a0dd73aa8e079d6373492a0c1cbcd70ed7082